### PR TITLE
Extract ExtraMeta symbolic shape fields into a dedicated SymbolicShap…

### PR DESCRIPTION
**Pull request by ppiskorski**
…eMeta

This modularizes ExtraMeta to bring down its creation cost when it is needed for other functions than syn shape handling.

Change-Id: Ife59b201b0c4fd75090fe8be5171a6dd73a10d10

Fixes #ISSUE_NUMBER

### CODE REVIEW

**Comment by ppiskorski** [comment]
> So, the enhanced version of the proposal is to refactor ExtraMeta so that it is generally extensible, and you don't have to pay the struct cost for pieces you are not using. With this proposal, we have the following "fields" in TensorImpl:
> 
>     * Symbolic sizes / strides (all the Sym fields)
> 
>     * NamedTensorMeta
> 
>     * BackendMeta
> 
> 
> So maybe the simplest form of the enhanced proposal is to move symbolic sizes / strides into a dedicated struct, and then just have three pointers in ExtraMeta to the various stuff. There is probably more fanciness (e.g., in case we want to add a lot more fields) but if you expose a method and not the struct directly we should be able to refactor it.

@ezyang this is a follow up on [the BackendMeta patch](https://github.com/pytorch/pytorch/pull/97429)  
pls have a look.

File: c10/core/TensorImpl.h [hunk 1]
```diff
@@ -2443,8 +2442,9 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 
   SymInt compute_sym_numel() const {
     TORCH_INTERNAL_ASSERT_DEBUG_ONLY(has_symbolic_sizes_strides_);
+    auto& sym_shape_meta{symbolic_shape_meta()};
```
Comment (ezyang) [inline c10/core/TensorImpl.h:2445]: weird but ok haha

File: c10/core/TensorImpl.h [hunk 2]
```diff
@@ -241,52 +241,37 @@ struct C10_API ExtraMeta {
...
-        numel_,
-        storage_offset_,
-        is_contiguous_,
-        is_channels_last_contiguous_,
-        is_channels_last_3d_contiguous_,
-        is_channels_last_,
-        is_channels_last_3d_,
-        is_non_overlapping_and_dense_,
-        named_tensor_meta_ ? named_tensor_meta_->clone() : nullptr,
-        backend_meta_ ? backend_meta_->clone(backend_meta_) : nullptr);
```
Comment (ezyang) [inline c10/core/TensorImpl.h:289]: nice

File: c10/core/TensorImpl.h [hunk 3]
```diff
@@ -241,52 +241,37 @@ struct C10_API ExtraMeta {
...
+  std::unique_ptr<c10::SymbolicShapeMeta> symbolic_shape_meta_ = nullptr;
   std::unique_ptr<c10::NamedTensorMetaInterface> named_tensor_meta_ = nullptr;
-  intrusive_ptr<c10::BackendMeta> backend_meta_;
+  intrusive_ptr<c10::BackendMeta> backend_meta_ = nullptr;
 
   ExtraMeta() = default;
+  ExtraMeta(const ExtraMeta& other) {
+    if (other.symbolic_shape_meta_) {
+      symbolic_shape_meta_ =
+          std::make_unique<c10::SymbolicShapeMeta>(*other.symbolic_shape_meta_);
```
Comment (ezyang) [inline c10/core/TensorImpl.h:255]: should we put a clone() method on SymbolicShapeMeta for this?

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**